### PR TITLE
ARROW-11659: [R] Preserve group_by .drop argument

### PR DESCRIPTION
**Pull request by ianmcook**
This preserves the `.drop` argument in the method for `dplyr::group_by()` and implements a method for `dplyr::group_by_drop_default()` which returns `TRUE` except on `arrow_dplyr_query` objects that were previously obtained by `group_by(.drop = FALSE)`.

### CODE REVIEW

File: r/tests/testthat/helper-data.R [hunk 1]
```diff
@@ -121,3 +121,15 @@ make_string_of_size <- function(size = 1) {
 
 example_with_extra_metadata <- example_with_metadata
 attributes(example_with_extra_metadata$b) <- list(lots = rep(make_string_of_size(1), 100))
+
+example_with_logical_factors <- tibble::tibble(
```
Comment (nealrichardson) [inline r/tests/testthat/helper-data.R:125]: 💯

Comment (ianmcook) [reply r/tests/testthat/helper-data.R:125]: 👊

File: r/R/dplyr.R [hunk 2]
```diff
@@ -42,7 +42,10 @@ arrow_dplyr_query <- function(.data) {
       filtered_rows = TRUE,
       # group_by_vars is a character vector of columns (as renamed)
       # in the data. They will be kept when data is pulled into R.
-      group_by_vars = character()
+      group_by_vars = character(),
+      # drop_empty_groups is a logical value indicating whether to drop
+      # groups formed by factor levels that don't appear in the data
+      drop_empty_groups = TRUE
```
[the remark this answers is not in the document]
Comment (ianmcook) [reply r/R/dplyr.R:48]: Yeah this was bothering me too that the default value of `TRUE` was set in multiple places like this. I will change it to `NULL` which is effectively the same as not setting it at all but I assume it's best to still show it here.

File: r/R/dplyr.R [hunk 3]
```diff
@@ -424,11 +427,16 @@ restore_dplyr_features <- function(df, query) {
   if (grouped) {
```
[the remark this answers is not in the document]
Comment (ianmcook) [reply r/R/dplyr.R:427]: I checked that, and no, it turns out that `dplyr::group_by(.drop = FALSE)` does the same thing as `dplyr::ungroup()`. It does not preserve `.drop`. One of the tests I added checks that we're consistent with dplyr on that: https://github.com/apache/arrow/pull/9716/files#diff-b54a9d9f7fc3cb71878c6396ac9b36893e7f60fb9f48c27fe41aeb5b8a1f5a48R237-R242

File: r/R/dplyr.R [hunk 4]
```diff
@@ -424,11 +427,16 @@ restore_dplyr_features <- function(df, query) {
...
     # Preserve groupings, if present
     if (is.data.frame(df)) {
-      df <- dplyr::grouped_df(df, dplyr::group_vars(query))
+      df <- dplyr::grouped_df(
+          df,
+          dplyr::group_vars(query),
+          drop = group_by_drop_default(query)
+        )
```
[the remark this answers is not in the document]
Comment (ianmcook) [reply r/R/dplyr.R]: Fixed in 5fc302fb84a7a9afc8fe7bf5d187716c88598e18

File: r/R/dplyr.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (ianmcook) [reply r/R/dplyr.R]: Changed in 5102fb9f6feccf8f62b6870728220d423e57e1cf

**Review by nealrichardson — APPROVED** [review]

**Pull request closed without merging**